Wires refactor

MMaiero opened the pull request:
This PR introduces a refactoring in the wires data model. Now a WireRecord will be composed by a Map<String,TypedValue<?>> that will ease the storage of a wireRecord in the database and the publishing to the Cloud Platform.
In general an emitter will emit a Wire Envelope composed by a List<WireRecord> but this list will contain only a WireRecord to simplify the storage and the publishing (a WireRecord is mapped to a Kura Payload).
The concepts of WireField and Severity Level are dropped.
Also ThrowableUtil class is deleted by this PR. Suitable code was used instead and only when needed.

The PR contains, also, a lot of bug fixes, cleanups and improvements.

MMaiero committed:
First drop of changes for the wires model refactor.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
New set of changes:

Removed severity level, added ErrorValue as new TypedValue.
Started changing components. Almost done for asset, timer, publisher.
Still to complete for the other components.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Improved the logging a bit.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Cleanup and reordered methods to meet Kura project classes structure.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Removed ErrorValue and WireField.

Now wire record contains only a Map object. Removed timestamp and position.
WireEnvelope contains a List of WireRecords.
Modified the related classes accordingly.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Solved issues in asset and component.provider.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Refactored db filter code. Various cleanups.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Few minor updates in wire filter, copyrights and wire database.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Modified logger to support two modalities:

INFO that is a limited logging; DEBUG that writes out all the content
of the received wire envelope.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Solved baseAsset issue that was causing the multiplication of
write requests to the driver.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Modified logger labels.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Deleted ThrowableUtil.java, updated doc, db improvements.

Deleted ThrowableUtil.java as not needed, changed code accordingly with new localized messages.
Started the updating of javadocs to meet the new architecture.
Improvements in db implementation to deal with blobs, solved issues on cleanup.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Cleanups.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Introduced localization caching to speedup overall behavior.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Solved timeout issue with openJDK.

With the openJDK the default timeout of 5000ms is not sufficient to have the services started and tracked.
Extended the timeout value to 60000 ms for both the emitter and receiver service trackers.
Some minor cleanups.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/WireRecord.java:
```diff
@@ -36,103 +33,67 @@
 @ThreadSafe
```
It's not thread safe as the map is mutable through `addProperty` method. You can apply GoF creational design patterns for this purpose. Have a look at the `WireRecord` implementation in my test_design branch.

amitjoy commented on kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/BaseAsset.java:
```diff
@@ -399,9 +365,8 @@ protected String getFactoryPid() {
-        return (oldAdId != ASSET_DESC_PROP.value()) && (oldAdId != ASSET_DRIVER_PROP.value())
-                && (oldAdId != NAME.value()) && (oldAdId != TYPE.value()) && (oldAdId != VALUE_TYPE.value())
-                && (oldAdId != SEVERITY_LEVEL.value());
+        return oldAdId != ASSET_DESC_PROP.value() && oldAdId != ASSET_DRIVER_PROP.value() && oldAdId != NAME.value()
```
I would still prefer clean separation of every predicate using `brackets` `()`.

MMaiero replied:
It's not in what we decided for our cleanup profile. Furthermore usually also Sonar complains for those brackets.

amitjoy replied:
Check the new comments. I don't think it has been specified in the formatter as I am also using the formatter but it still doesn't show me any problem. Apart from it, there are also few places in this PR, where the braces remain as well.

MMaiero replied:
The cleanup profile is not the formatter

MMaiero replied:
Please tell me where I missed them

amitjoy replied:
will comment in further revisions of this review 👍

MMaiero commented:
Closes #1057 
Closes #1056 
Closes #1041

amitjoy commented on kura/org.eclipse.kura.localization/src/main/java/org/eclipse/kura/localization/LocalizationAdapter.java:
```diff
@@ -38,18 +46,45 @@
+    private static class C10NWrapper<T> implements InvocationHandler {
+
+        private T wrapped;
+        private HashMap<Method, String> cache = new HashMap<>();
```
make both of these states as final

amitjoy commented on kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/base/ThrowableUtil.java:
```diff
@@ -1,54 +0,0 @@
- * The Class ThrowableUtil contains all necessary static factory methods for
- * manipulating Throwable instances
- */
-public final class ThrowableUtil {
```
Don't drop it right away. I have written an important and more useful method which I am making use of in other locations. In addition, instead of continuing to use `org.eclipse.kura.core.util.ThrowableUtil`, we can use this.

amitjoy replied:
I will make the necessary changes to this class after this gets merged.

MMaiero replied:
Currently not needed. Eventually will be added if needed later on

amitjoy replied:
Due to repeating the same amount code in different places, I would recommend this to conform to DRY principle.

amitjoy commented on kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceUtil.java:
```diff
@@ -61,7 +62,10 @@ private ServiceUtil() {
-            throw new KuraRuntimeException(KuraErrorCode.INTERNAL_ERROR, ThrowableUtil.stackTraceAsString(ise));
+            final StringWriter sw = new StringWriter();
+            final PrintWriter pw = new PrintWriter(sw);
+            ise.printStackTrace(pw);
```
As I mentioned earlier, instead of using org.eclipse.kura.core.util.ThrowableUtil.stackTraceAsString or writing it all again is not necessary. You still can use the utility methods from ThrowableUtil inside Utility bundle.

MMaiero replied:
Since it is used only few times, it is not needed to have ThrowableUtil

amitjoy commented on kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceUtil.java:
```diff
@@ -61,7 +62,10 @@ private ServiceUtil() {
+            final StringWriter sw = new StringWriter();
+            final PrintWriter pw = new PrintWriter(sw);
+            ise.printStackTrace(pw);
+            throw new KuraRuntimeException(KuraErrorCode.INTERNAL_ERROR, sw.toString());
```
Use any other suitable error code instead of INTERNAL_ERROR

MMaiero replied:
Agreed

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/filter/WireRecordCache.java:
```diff
@@ -1,163 +0,0 @@
- * The Class WireRecordCache is responsible to contain the Wire Record cached
- * values.
- */
-final class WireRecordCache {
```
Any specific reason of removing cache mechanism for DB Filters?

MMaiero replied:
The cache mechanism exists, it is provided in a simplified way.

amitjoy replied:
Just noticed 👍

amitjoy commented on kura/org.eclipse.kura.wire.provider/src/main/java/org/eclipse/kura/internal/wire/WireServiceImpl.java:
```diff
@@ -52,126 +57,167 @@
      */
-    synchronized void createWires() throws KuraException {
-        s_logger.debug(s_message.creatingWires());
+    synchronized void createWires() {  // TODO: this needs to be refactored
```
If you want to refactor, you can do it with this PR.

MMaiero replied:
Timing issues. All the todos are tracked and will be solved in M2. For now these methods are taken as they are.

amitjoy replied:
I am curious to know more about it

amitjoy commented on kura/org.eclipse.kura.wire.provider/src/main/java/org/eclipse/kura/internal/wire/WireServiceImpl.java:
```diff
@@ -265,9 +297,12 @@ public void deleteWireConfiguration(final WireConfiguration wireConfiguration) {
+            final StringWriter sw = new StringWriter();
+            final PrintWriter pw = new PrintWriter(sw);
+            e.printStackTrace(pw);
+            throw new KuraRuntimeException(KuraErrorCode.INTERNAL_ERROR, sw.toString());
```
DRY - Do not repeat yourself. I would still prefer ThrowableUtil to be used. Do not drop it. It is proven useful.

amitjoy reviewed (changes requested):
v1.0 comments done.

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/subscriber/CloudSubscriber.java:
```diff
@@ -71,32 +74,32 @@
 
         @Override
-        public void removedService(ServiceReference<CloudService> reference, CloudService service) {
+        public void removedService(final ServiceReference<CloudService> reference, final CloudService service) {
```
Make this consistent. Sometimes, the final modifiers are removed and sometimes it is added.

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/subscriber/CloudSubscriber.java:
```diff
@@ -356,67 +355,59 @@ private void setupCloudClient() throws KuraException {
+                final byte[] value = TypeUtil.objectToByteArray(entryValue);
+                convertedValue = TypedValues.newByteArrayValue(value);
+            } else {
+                throw new IllegalArgumentException("Unknown metric type");
```
Add this literal to localization resource

amitjoy commented on kura/org.eclipse.kura.wire.provider/src/main/java/org/eclipse/kura/internal/wire/WireServiceImpl.java:
```diff
@@ -279,43 +314,43 @@ public void deleteWireConfiguration(final WireConfiguration wireConfiguration) {
             final String key = entry.getKey();
-            final boolean isNotWireConfigurationProperty = !(key.endsWith(s_message.emitter())
-                    || key.endsWith(s_message.receiver()) || key.endsWith(s_message.filter()));
+            final boolean isNotWireConfigurationProperty = !(key.endsWith(message.emitter())
```
I don't think it's a formatter issue but an user-specific configuration to use or not use any brackets.

amitjoy reviewed (changes requested):
v1.2 comments done

amitjoy commented on kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtWireServiceImpl.java:
```diff
@@ -71,22 +71,21 @@
  */
 public final class GwtWireServiceImpl extends OsgiRemoteServiceServlet implements GwtWireService {
 
+    private static final int SERVICE_WAIT_TIMEOUT = 60000;
```
Is there any specific reason of setting it to 1 min? For embedded devices, I don't think it's good idea to wait for this much amount of time.

MMaiero replied:
The reason is that openJDK is a way slower than the Oracle one. Currently, with 5 seconds that would imply an Exception due to the fact that the wire creation was called with services instances still not running.

amitjoy reviewed (commented):
v1.3 comments

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/store/DbWireRecordStore.java:
```diff
@@ -171,29 +236,30 @@ public synchronized void bindWireHelperService(final WireHelperService wireHelpe
     private void clear(final int noOfRecordsToKeep) {
-        final String sqlTableName = this.dbHelper.sanitizeSqlTableAndColumnName(this.options.getTableName());
+        final String sqlTableName = this.dbHelper
+                .sanitizeSqlTableAndColumnName(this.wireRecordStoreOptions.getTableName());
```
extract the table name to a local var

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/store/DbWireRecordStore.java:
```diff
@@ -171,29 +236,30 @@ public synchronized void bindWireHelperService(final WireHelperService wireHelpe
             final String catalog = conn.getCatalog();
             final DatabaseMetaData dbMetaData = conn.getMetaData();
-            final ResultSet rsTbls = dbMetaData.getTables(catalog, null, sqlTableName, null);
+            final ResultSet rsTbls = dbMetaData.getTables(catalog, null, this.wireRecordStoreOptions.getTableName(),
```
use local var instead

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/store/DbWireRecordStore.java:
```diff
@@ -395,124 +414,120 @@ private void reconcileColumns(final String tableName, final WireRecord wireRecor
+            final Object value = entry.getValue();
+            switch (dataType) {
+            case BOOLEAN:
+                logger.debug(message.storeBoolean(((BooleanValue) value).getValue()));
```
remove all the debug statements as these are more verbose and hence removed from DB Filter as well

amitjoy commented on kura/org.eclipse.kura.wire.helper.provider/src/main/java/org/eclipse/kura/internal/wire/helper/WireHelperServiceImpl.java:
```diff
@@ -113,16 +123,14 @@ public String getServicePid(final WireComponent wireComponent) {
-                        return true;
-                    }
+                final WireComponent wc = (WireComponent) context.getService(ref);
+                if (ref.getProperty(KURA_SERVICE_PID).equals(wireComponentPid) && wc instanceof WireEmitter) {
```
getting service instance is more expensive than checking equals on string instance. It should be :

1. Check the property from the service reference
2. If true, get the service instance
3. Check if Emitter
4. If true, return the instance

amitjoy commented on kura/org.eclipse.kura.wire.helper.provider/src/main/java/org/eclipse/kura/internal/wire/helper/WireHelperServiceImpl.java:
```diff
@@ -134,16 +142,14 @@ public boolean isEmitter(final String wireComponentPid) {
-                    if (wc instanceof WireReceiver) {
-                        return true;
-                    }
+                final WireComponent wc = (WireComponent) context.getService(ref);
```
Same here

amitjoy reviewed (changes requested):
v1.4 Comments

amitjoy commented on kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtWireServiceImpl.java:
```diff
@@ -71,22 +71,21 @@
     private static final String PRODUCER = "producer";
-    /** The Logger instance. */
+
     private static final Logger s_logger = LoggerFactory.getLogger(GwtWireServiceImpl.class);
```
Kindly change it to `logger`

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/asset/WireAsset.java:
```diff
@@ -59,47 +60,88 @@
  * be written to {@code B}'s channel
  *
+ * @see Channel
+ * @see AssetRecord
```
Add `DriverRecord` to it as well

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/asset/WireAsset.java:
```diff
@@ -179,155 +254,60 @@ protected synchronized void deactivate(final ComponentContext context) {
      * @return list of Asset Records containing the values to be written
      * @throws NullPointerException
      *             if argument is null
      * @throws IllegalArgumentException
```
remove `IllegalArgumentException`

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/asset/WireAsset.java:
```diff
@@ -179,155 +254,60 @@ protected synchronized void deactivate(final ComponentContext context) {
-            throw new IllegalArgumentException(s_message.wireRecordsNonEmpty());
-        }
+    private List<AssetRecord> determineWritingChannels(final WireRecord record) {
+        requireNonNull(record, message.wireRecordsNonNull());
```
change it to `wireRecordNonNull()` and change the corresponding localization string literal

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/filter/DbWireRecordFilter.java:
```diff
@@ -286,69 +258,58 @@ public void producersConnected(final Wire[] wires) {
     /**
-     * Schedule refresh of SQL view operation
+     * Trigger data emit as soon as new {@link WireEnvelope} is received. The component caches the last database
+     * read and
```
reorganize the comment. Whitespaces are scattered.

amitjoy commented on kura/org.eclipse.kura.wire.component.provider/src/main/java/org/eclipse/kura/internal/wire/publisher/CloudPublisher.java:
```diff
@@ -431,15 +364,16 @@ private void publish(final List<WireRecord> wireRecords) {
             for (final WireRecord dataRecord : wireRecords) {
```
As discussed previously, we should publish all the collected `WireRecords` all at once.

amitjoy replied:
I beg your pardon. It's not possible. We have to send every `WireRecord` per payload. 👍

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/WireRecord.java:
```diff
@@ -36,103 +32,28 @@
-        this.timestamp = timestamp;
-        this.position = null;
-        this.fields = Collections.unmodifiableList(fields);
+        this.properties = properties;
```
Make it immutable. `new HashMap<>(properties)`. Otherwise clients can still be able to mutate the state of `WireRecord` using the provided map reference.

amitjoy replied:
or `Collections.unmodifiableMap(properties)` but this is a costly operation.

amitjoy reviewed (changes requested):
v1.5 comments

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/WireRecord.java:
```diff
@@ -36,103 +33,28 @@
 @ThreadSafe
 public class WireRecord {
```
Also add `@Immutable`

MMaiero replied:
look better at the code.
It is @immutable

amitjoy reviewed (changes requested):
v1.6 RC 😉 comments
Could you kindly solve the build issues as done in #1046?

MMaiero committed:
Extracted payloadType as variable in CloudPublisher's publish method.

Other minor fixes.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented:
Let me have a final look.

amitjoy commented on kura/org.eclipse.kura.web2/src/main/java/org/eclipse/kura/web/server/GwtWireServiceImpl.java:
```diff
@@ -495,19 +495,21 @@ public GwtWiresConfiguration updateWireConfiguration(final GwtXSRFToken xsrfToke
 
                     wireService.createWireConfiguration(emitterPid, receiverPid);
+
+                    producerTracker.close();
```
Move it before creating wire configuration because if the wire configuration creation fails, the tracker will never be closed.

MMaiero replied:
I'm still working on this. Currently, Kura with openJDK is nearly unusable.

amitjoy replied:
Let's do it this way. Have the trackers closed before creating the wire configuration and then we can merge this. You can continue working on resolving the OpenJDK issue thereafter.

amitjoy commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/type/TypedValues.java:
```diff
@@ -19,8 +19,10 @@
+ * @noextend This class is not intended to be extended by clients.
  */
-public final class TypedValues {
+public class TypedValues {
```
As this is an utility class, I strongly prefer to have final as consumers can extend this by adding new methods to it without breaking any functionalities. As there won't be any instances of this class, it will not be extended by other classes. Such extensions for utility classes are not at all common in Java community.

MMaiero committed:
Changes to typedValues.

Added static method newTypedValue that tries to map a received object with the corresponding TypedValue.
Cleaned up, correspondingly, DbWireRecordFilter.java and CloudSubscriber.java.
Moved up tracker closes in GwtWireServiceImpl.java.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero commented:
Ehm, maybe it was better to wait before merging. Anyway, let's leave it as is...
